### DISCUSSION TITLE
Add support for Nostr NIP-57 Lightning Zaps to LN Address / LNURL

### DISCUSSION THREAD

**Discussion by derekross** [btcpayserver/btcpayserver#4642] [Ideas / Feature Requests]
See https://github.com/nostr-protocol/nips/blob/master/57.md for more information. Thank you.

**Comment by mp21-org** [comment]
👍 for this feature request

**Comment by dennisreimann** [comment]
Started to look into this today, also reached out to @Kukks to coordinate. :)

**Comment by mp21-org** [comment]
Loosly related: Could we also get the LNURL comment (i.e. the zapping comment that the user added) as email variable here:
https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer/HostedServices/StoreEmailRuleProcessorSender.cs (towards the end)

**Reply by Perlover** [reply]
And please don't forget LNBank + Lightning Address bundle - please check it too

**Comment by JamesTsetsekas** [comment]
I'm also interested in this feature!

**Comment by mp21-org** [comment]
![grafik](https://user-images.githubusercontent.com/1374810/219867697-dfebfc90-4204-4bea-9f9f-aeb32c14bd36.png)

**Comment by michaelhall923** [comment]
+1 for this

**Comment by sethforprivacy** [comment]
Will contribute 250,000 sats to a bounty for this feature, to be paid to the proposer of the PR that implements it.

Desperately need non-custodial Lightning Address + NIP-57 support before too many people jump ship and use custodial wallets.

**Reply by InnocuousFinch** [reply]
I'm in for 50k sats

**Reply by ttiiggss** [reply]
I'm in for 50k too.

**Reply by 0xtrr** [reply]
I'll pledge an additional 100k sats

**Reply by michaelhall923** [reply]
Put me in for 50k

**Reply by v0l** [reply]
Adding another 250k

**Reply by BitcoinQnA** [reply]
I'm in for 100k sats

**Reply by sethforprivacy** [reply]
Thanks for all the added support, everyone! Love seeing this grow, shows that there is genuine support and desire for powerful non-custodial tools for LN + Nostr.

**Reply by TiKawamoto** [reply]
I'm in for 300k

**Reply by sethforprivacy** [reply]
@InnocuousFinch @ttiiggss @0xtrr @michaelhall923 @v0l @BitcoinQnA @TiKawamoto please payout your pledge via https://github.com/btcpayserver/btcpayserver/discussions/4642#discussioncomment-5543650 and post proof under that thread.

Thank you all for your contributions, love to see the Bitcoin community coming together to support open-source devs and incentivize important features to be built!

**Comment by derekross** [comment]
~~We're up to 850K bounty now on this.~~

~~We're up to 1.4M now including the sats from Nostr zaps that I will be donating as well.~~

We're up to approximately 1.7M sats now.

**Reply by bitcoinlizard** [reply]
Add 100k sats to the bounty for me.

**Reply by derekross** [reply]
@bitcoinlizard it's time to pay up :) the feature has been implemented!

**Comment by mutatrum** [comment]
Maybe this can help for code inspiration: https://github.com/Dolu89/ligess/pull/9/commits/942fcad36b418a4cf87d09b40bf27fb7f87f062b#diff-8b6041d688f535453a00cedbb54b1cb4b3b20abd6bec228a797731b8073d0bfe

I don't have a C# environment, but the gist of validation and construcing the zap note are easy to read.

**Comment by v0l** [comment]
Strike.army already implements Zaps in C# using NNostr and BtcPayServer libs from Nuget, You can probably easily copy it from here:

https://github.com/v0l/strike.army/blob/main/src/Controllers/PayController.cs
https://github.com/v0l/strike.army/blob/main/src/Services/ZapService.cs

**Comment by utxo-one** [comment]
Put me down for 300k sats and who I should pay

**Reply by sethforprivacy** [reply]
We'll pay directly to the person who implements the feature via a PR so that no one person custodies the entire bounty.

**Reply by sethforprivacy** [reply]
@utxo-one please payout your pledge via https://github.com/btcpayserver/btcpayserver/discussions/4642#discussioncomment-5543650 and post proof under that thread.

**Comment by ok300** [comment]
Have you considered that it's opening up the BTCPayServer to DDoS?

Creating a zap request is free. Upon a request, the receiver has to create an invoice and monitor it until it's paid, in order to send the zap note.

In other words, with one free action, anyone from the internet can cause a BTCPay to expend resources for a few minutes (memory and CPU cyles) until the invoice is paid or expired.

Its trivial to write a script that creates endless zap requests.

**Reply by sethforprivacy** [reply]
Isn't this always true of invoice systems like BTCPay?

You can do the same with BTCPay invoices currently, and all are monitored for 10min until they expire.

If Zaps are treated the same way I don't see an issue, and it's up to the person running BTCPay if they enable this feature or not anyways.

**Reply by ok300** [reply]
Not necessarily, since zaps and nostr are a different network with different properties.

For the HTTP part, BTCPay has rate limiting (#3889) to prevent DDoS. The LN node it uses also has similar defenses against bad actors on the LN network. There are other defenses available too: intermediaries like Cloudflare, or only allowing requests from authenticated clients, etc.

However it's more tricky with zaps. How would you know if a zap request is honest (will pay the invoice) or dishonest (attacker)? There is no source IP (there are source pubkeys, but its trivial to create new ones). You cannot block a relay, because the attacker can use multiple relays.

The only option I can think of is to rate-limit "the total amount of new zaps accepted per minute". But this only opens the door for a different kind of DoS, where the attacker doesn't care about taking down your service, but only wants to deny others the ability to zap you. They can issue a flood of new zaps until BTCPay reaches the zap rate-limit. At that point, you cannot receive any zaps from anyone.

So, with a simple script, anyone in the world can cause any zap receiver to stop receiving zaps.

Or am I missing something?

Is there a plausible defense against this?

**Reply by mutatrum** [reply]
Counter-intuitively, the zap request is not going over Nostr, the zap request is embedded in the LN-URL HTTP call. So the same defences should be in place for those.

**Reply by ok300** [reply]
You're right! Good catch.

**Reply by brianoflondon** [reply]
> So, with a simple script, anyone in the world can cause any zap receiver to stop receiving zaps.

This has been my nervousness with LNURLp ever since I implemented it.

> Is there a plausible defense against this?

I really need to look at rate limiting and defence like that.

**Comment by damanic** [comment]
+1 🙏

**Comment by mutatrum** [comment]
Ok, I started. Please let me know if someone else already did something.

**Comment by Kukks** [comment]
Please do it as a plugin if you do build it. Imo this should not be a core
feature.

**Comment by mutatrum** [comment]
I'm not sure how feasible separating it into a plugin is. Lightning Address support is sort of a core feature as well as a bundled plugin (?) and to support zaps, that logic needs to be adjusted a fair bit, as zap requests needs specialised handling to construct the invoice.

**Reply by dennisreimann** [reply]
Nostr functionality shouldn't be a part of BTCPay Server Core, especially not since we introduced the plugin system. Yes, this will also involve discussion on how to best achieve that, addition of hooks, etc. — if it would be easy, this would have been implemented already, but we decided not to rush it and get it done properly if at all.

**Comment by Kukks** [comment]
Zap support has been implemented in the ~NIP05~Nostr plugin, and the new version will be released alongside 1.9 🎊

Please donate all bounties over at https://opensats.org/projects/btcpayserver and post a screenshot as proof. I'll let @derekross / @sethforprivacy handle coordination as the bounty starters

**Reply by sethforprivacy** [reply]
Holy shit, love it! Sending my pledge now.

**Reply by Kukks** [reply]
Maybe wait till it's out so that there is no rugpullity 😁

**Reply by sethforprivacy** [reply]
NO TIME TO LOSE:

![image](https://user-images.githubusercontent.com/40500387/230393032-430798cd-2527-477c-b702-b45d46facb13.png)

**Reply by sethforprivacy** [reply]
For other's paying out their pledge to the bounty, note that OpenSats doesn't let you select a satoshi amount.

Can use a converter like this to get a rough fiat amount and then use that on OpenSats: https://www.kraken.com/learn/satoshi-to-usd-converter

**Reply by derekross** [reply]
AWESOME! This is truly exciting. I will be paying all of the bounties from note12gg3s27h7w4vtpdzttszx5u4jz96ddlny2qe8g3ac3t0hwkpml4sl3m03l shortly.

**Reply by Kukks** [reply]
> For other's paying out their pledge to the bounty, note that OpenSats doesn't let you select a satoshi amount.
> 
> Can use a converter like this to get a rough fiat amount and then use that on OpenSats: [kraken.com/learn/satoshi-to-usd-converter](https://www.kraken.com/learn/satoshi-to-usd-converter)

wtf! I'll have to do a PR to allow sats on there :)

**Reply by 0xtrr** [reply]
> NO TIME TO LOSE:
> 
> ![image](https://user-images.githubusercontent.com/40500387/230393032-430798cd-2527-477c-b702-b45d46facb13.png)

Did you pay over LN? 👀 
I went to the link @Kukks posted and was only offered to pay on-chain, which I did😅

**Reply by sethforprivacy** [reply]
Mentioned all users who pledged and requested they payout via your directions, @Kukks. Thanks again *so much* for building this feature out, can't wait to update to 1.9 and see the Zaps flow :zap:

**Reply by sethforprivacy** [reply]
> > NO TIME TO LOSE:
> 
> Did you pay over LN? eyes I went to the link @Kukks posted and was only offered to pay on-chain, which I didsweat_smile

The first time I clicked it without selecting an amount it only provided on-chain, second time I provided a fiat amount and it offered LN. Might be a bug or they might be having some LN node issues?

**Reply by 0xtrr** [reply]
> > > NO TIME TO LOSE:
> > 
> > 
> > Did you pay over LN? eyes I went to the link @Kukks posted and was only offered to pay on-chain, which I didsweat_smile
> 
> The first time I clicked it without selecting an amount it only provided on-chain, second time I provided a fiat amount and it offered LN. Might be a bug or they might be having some LN node issues?

Ah ok, sounds like a bug to me. I'd honestly rather pay over LN if that was possible. Oh well, my TX is having fun with all the other bitcoin NFT transactions😂

**Reply by derekross** [reply]
Paid! 
![image](https://user-images.githubusercontent.com/755144/230397286-dfe064eb-42f5-4a69-b75f-a49268b9897e.png)

**Reply by TiKawamoto** [reply]
Looking forward to this!

![image](https://user-images.githubusercontent.com/3180203/230397168-cc3db173-e8aa-46f8-99a6-ab62ea5d4455.png)

**Reply by Kukks** [reply]
> > > NO TIME TO LOSE:
> > 
> > 
> > Did you pay over LN? eyes I went to the link @Kukks posted and was only offered to pay on-chain, which I didsweat_smile
> 
> The first time I clicked it without selecting an amount it only provided on-chain, second time I provided a fiat amount and it offered LN. Might be a bug or they might be having some LN node issues?

https://github.com/OpenSats/website/issues/42
https://github.com/OpenSats/website/issues/41

**Reply by utxo-one** [reply]
Thanks for building this, will test it out this weekend.

![image](https://user-images.githubusercontent.com/111649294/230399471-606a36d7-bbf6-4c42-9fe5-7d5fa64ffcdb.png)

**Reply by InnocuousFinch** [reply]
Cheers @Kukks ![image](https://user-images.githubusercontent.com/118941419/230402915-1b41f223-d2d3-4b09-b9bc-ff6c8159d167.jpeg)

**Reply by michaelhall923** [reply]
Thanks so much! 
![Receipt from OpenSats](https://user-images.githubusercontent.com/40177960/230648421-47d9e6fc-511d-491d-92e5-b98ea193b7a2.jpg)

**Reply by sethforprivacy** [reply]
> Zap support has been implemented in the ~NIP05~Nostr plugin, and the new version will be released alongside 1.9 confetti_ball
> 
> Please donate all bounties over at https://opensats.org/projects/btcpayserver and post a screenshot as proof. I'll let @derekross / @sethforprivacy handle coordination as the bounty starters

Hey Kukks, just saw 1.9.0 got released and I updated, but I only see the old 1.0.2 NIP-05 plugin that doesn't seem to have Zap support still. Do I need to do anything specific to get the latest version of that plugin?

**Reply by mp21-org** [reply]
Server Settings - Policies - Plugins:
![grafik](https://user-images.githubusercontent.com/1374810/231761756-a5b4a294-5e08-40c4-8df5-ceb45bd4da27.png)

That will yield the newer plugin as an update under manage plugins. Not sure when it will come out of pre-release status. @Kukks can for sure say more.

**Comment by 0xtrr** [comment]
Appreciate this contribution @Kukks!
![fdsfdsfdsfsd](https://user-images.githubusercontent.com/86188777/230421241-5b608b62-94f7-4ecb-a10f-84920ab2592c.png)

**Comment by bitcoinlizard** [comment]
Great job @Kukks !
![image](https://user-images.githubusercontent.com/75000074/230454507-b3fca299-b912-46a9-a577-e218869a57bf.png)

**Reply by Perlover** [reply]
Thanks!

https://btcpay0.voltageapp.io/i/BpUnSdKKpwbZRNVdfFcGg7/receipt

**Comment by v0l** [comment]
Paid @Kukks 
![Screenshot 2023-04-06 at 22 21 56](https://user-images.githubusercontent.com/1172179/230496504-951d4eab-25c4-45bc-b964-ea65ca5dab21.png)

**Comment by BitcoinQnA** [comment]
Paid, thank you!

![opensats](https://user-images.githubusercontent.com/64638670/231086231-7c7ed0aa-a9f2-4aaf-90d2-ba602afa8fdd.png)

**Comment by Chiliss** [comment]
Y a ora k procede

**Comment by Sjors** [comment]
I'm really confused what the blue message is trying to say:

<img width="880" alt="Scherm­afbeelding 2023-09-05 om 21 06 09" src="https://github.com/btcpayserver/btcpayserver/assets/10217/f92bf2b5-16ed-4b90-a824-913c5adacef9">

Do I need to delete the plugin?

In any case people can zap me, but the zaps don't show up. Adding an nsec (or a relay) to the form just disappears after submit.

**Reply by dennisreimann** [reply]
@Sjors Yes, you'll need to keep the plugin active, but for general zap functionality only a Lightning Address is needed. However, I'm seeing the same thing, @Kukks: The zaps don't show up in the clients. They get sent and end up on the node, also the info in BTCPay looks just fine, but Nostr clients just don't show them.

**Reply by Perlover** [reply]
> @Sjors Yes, you'll need to keep the plugin active, but for general zap functionality only a Lightning Address is needed. However, I'm seeing the same thing, @Kukks: The zaps don't show up in the clients. They get sent and end up on the node, also the info in BTCPay looks just fine, but Nostr clients just don't show them.

Maybe clients don't show zaps simply because the plugin doesn't work correctly? After all, customers are guided by special receipts within the Nostr network, which are issued after payment. To do this, the plugin must generate a receipt after receiving the zap and send it via relay to the Nostr network.

Is it doing this? When I tested this plugin, I did not observe anything like this, at least as a regular user.

**Reply by Kukks** [reply]
Basically, it's been a series of unfortunate events with various combinations of btcpay version, lnurl lib version, nostr lib version and the nostr plugin. 

Since we use a compiled language and then have a plugin system that loads at runtime, common libraries are loaded once and conflicting versions are on a best effort situation causing silent breakages. I am working on making the plugin system a bit more solid, or at least provide more alerts if something breaks. 

BTCPay also went through a bug refactor for ln addresses so that we handle scalability a lot better (we generated an invoice for every ping which was basically a  ddos vector when combined with nostr zaps), which then broke the way we detect zaps. 


Ill leave an update here soon once I give this a test and let you know the end result. In the last btcpay version and nostr plugin, I thought I got all the plumbing correctly, but may have missed something.

**Reply by Sjors** [reply]
I'll update to v1.11.4 now, since I was on v1.11.3. Will also hit update on the plugin if there is such a button.

**Reply by Sjors** [reply]
By the way I prefer butal unforgving error messages over best effort managing :-)

**Reply by Sjors** [reply]
I see the plugin was at `1.0.13`,upgrading it now.

**Comment by derekross** [comment]
I'll chime in and say me too. I could never get them to work many months
ago when this feature launched. The Lightning payment works, but the Zap,
the callback to Nostr to sign the event, never worked, and clients never
saw the Zap. I have not tried since the original launch.

Derek Ross
https://NostrPlebs.com
https://NostrPlebs.com/s/derekross

**Comment by Kukks** [comment]
@derekross  @dennisreimann  @Perlover @Sjors 


I just pushed nostr 1.1.0 plugin nd used it alongside BTCPay 1.11.4.

Now, every lnurl generated by btcpay supports zaps. you can have your nostr profile pointing to a pay button lnurl and it will be zappable. 
When an  invoice generated for a zap is paid, the invoice details page will show which relays it will broadcast to, and what the zap request and receipt ids are. 

I believe this should be fixed once and for all!

**Reply by Sjors** [reply]
Updated! What's a "pay button lnurl"? I have the following redirect:
<img width="587" alt="Scherm­afbeelding 2023-09-07 om 09 44 05" src="https://github.com/btcpayserver/btcpayserver/assets/10217/0495694f-4035-476c-ad3a-37d7c8f4c5cc">

And then in the Lightning Address plugin I have: `sjors@pay.sprovoost.nl`.

**Reply by Sjors** [reply]
I just tried zapping myself using the Phoenix wallet. The invoice settles, but nothing seems to get broadcast.

**Reply by Sjors** [reply]
Screenshot of the log:
<img width="723" alt="zap" src="https://github.com/btcpayserver/btcpayserver/assets/10217/213b0aa1-7ab3-42b3-a2ee-42a64c47e8ee">

**Reply by stupleb** [reply]
Same here. Zaps continue to be collected, but no subsequent events received by relays.

I also have my own relay, so I can use that to monitor. These are the logs from the relay...

```
2023-09-14T16:42:00+02:00  Sep 14 14:42:00.074  INFO nostr_rs_relay::server: new client connection (cid: 6a91a1ae, ip: "172.18.0.1")
2023-09-14T16:42:01+02:00  Sep 14 14:42:01.105  INFO nostr_rs_relay::server: stopping client connection (cid: 6a91a1ae, ip: "172.18.0.1", sent: 0 events, recv: 0 events, connected: 1.031439626s)
```

This shows BTCPay connecting then doing nothing and disconnected in response to a test zap.

**Reply by stupleb** [reply]
I think I've got it. It looks like the plugin converts an npub to a hex incorrectly, and then publishes with that in the metadata (and in my case fails to publish on my local relay, but that's another issue). Add your correct hex instead of the npub, and it's working. cc @Sjors

**Reply by Sjors** [reply]
@stupleb I'll try that. Do I also need to add relays or does BTCPay has some defaults? Last time I tried to add any they just disappeared after submitting the form.

**Reply by Sjors** [reply]
Actually that's not it for me, BTCPay had the correct hex, assuming this site works: https://nostrcheck.me/converter/

**Reply by Sjors** [reply]
Also how is it supposed to sign without having my nsec?

**Reply by Sjors** [reply]
Ah, this time it _does_ store the relays I add. And that seems to have fixed the zap notifications!

Ok, so the message "None of the nostr settings on this page are needed for zaps anymore." is fake news.

**Reply by stupleb** [reply]
You seem to have got it but I'l answer anyway...

Yes, add the hex of your private key too (in the right place, for anyone reading!!). Add relays with wss:// too.

Regarding your comment on the correct hex, I looked in the metadata of the event sent to relays and the pubkey was wrong. Don't know how or why. But switching from npub to hex and the following events had the correct pubkey in the events.

**Reply by Sjors** [reply]
No I missed the private key hex part. Have to figure out how to convert that on the command line.

**Reply by Sjors** [reply]
When I enter a hex value under "Private key for zaps" it just disappears after submitting. However it does seem to work now... though last time I saw a zap from myself and while later couldn't find it anymore.

**Reply by Kukks** [reply]
@sjors Can confirm that the private key field was not being populated due to some weirdness in the UI code.

1.1.1 is out with a fix for this! I've also removed my batching mechanism which was meant to make things less wasteful (not opening a websocket connection to every relay for every individual zap) because I'm 169% certain this is what has been causing confusion around "zaps not working". 

This is the relevant code block which is gone now:
![image](https://github.com/btcpayserver/btcpayserver/assets/1818366/5d351357-a5d3-41d9-9dec-73ffde406f12)


Instead, it will now process pending zaps every 5 seconds

**Reply by Kukks** [reply]
This is impossible. When you put in npub/nsec in the UI, I instantly convert to the hex variant and save that instead.

**Reply by Sjors** [reply]
Ok, I'll try running 163f805f3b922b4030e0a6cbc720f9f14c6cf5d4 for a bit.

**Reply by Sjors** [reply]
I don't see a difference in behavior when saving an nsec: it still disappears from the UI.

**Comment by bitcoinlizard** [comment]
For what it's worth, this seems to be working correctly for me.  I do the NIP-5 stuff myself with my own web server, not with the btcpay server plugin.  I have a lightning address configured and the nostr plugin installed and indeed, as the plugin configuration screen says, I don't need to fill in any of the information.  It just works.

This worked correctly for me with plugin version 1.1.0 and I just upgraded and it still works with 1.1.1.  As any nostr user I can zap a post or my profile and the zap will show up.  I can see on the paid invoice in btcpay server under "Additional Information" that my zap was broadcast so several relays although it seems like these relays must be hard coded as they are not the relays configured for my npub.

If you want to see this working in action feel free to zap me :)

https://snort.social/p/npub1k6f0ecam577lay8mm8azfvs3u4xvl23q82zpc3htee6ath09gg0s9ftryn

**Reply by dennisreimann** [reply]
Can confirm it works for me too now. I assume the two minute delay was a part of the confusion.

**Comment by Kukks** [comment]
The relay list comes from a combination of the relays you have listed and
the relays the zap request event specifies.

Thank you for feedback, needed some positivity in here 😭

**Comment by Sjors** [comment]
Ah, the Additional Information is useful.
